src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs

```diff
@@ -988,29 +1034,87 @@ public virtual TokenValidationResult ValidateToken(string token, TokenValidation
-                    };
+                    // Set current configuration as LKG if it exists.
+                    if (validationParametersCopy.Configuration != null && validationParametersCopy.Configuration != validationParametersCopy.ConfigurationManager.LastKnownGoodConfiguration)
+                        validationParametersCopy.ConfigurationManager.LastKnownGoodConfiguration = validationParametersCopy.Configuration;
+
+                    return tokenValidationResult;
                 }
-                else
+                else if (tokenValidationResult.Exception.GetType().Equals(typeof(SecurityTokenInvalidSignatureException))
+                   || tokenValidationResult.Exception.GetType().Equals(typeof(SecurityTokenInvalidSigningKeyException))
+                   || tokenValidationResult.Exception.GetType().Equals(typeof(SecurityTokenInvalidIssuerException))
+                   || tokenValidationResult.Exception.GetType().Equals(typeof(SecurityTokenUnableToValidateException))
+                   || tokenValidationResult.Exception.GetType().Equals(typeof(SecurityTokenSignatureKeyNotFoundException)))
                 {
-                    var jsonWebToken = ValidateSignature(token, validationParametersCopy);
-                    return ValidateTokenPayload(jsonWebToken, validationParametersCopy);
+                    if (validationParametersCopy.ConfigurationManager.UseLastKnownGoodConfiguration
+                        && validationParametersCopy.ConfigurationManager.LastKnownGoodConfiguration != null
+                        && !(validationParametersCopy.ConfigurationManager.LastKnownGoodConfiguration == validationParametersCopy.Configuration))
+                    {
+                        // Inform the user that the LKG is expired.
+                        if (!validationParameters.ConfigurationManager.IsLastKnownGoodValid)
+                            LogHelper.LogInformation(TokenLogMessages.IDX10262);
+                        else
+                        {
+                            validationParametersCopy.Configuration = validationParametersCopy.ConfigurationManager.LastKnownGoodConfiguration;
+                            tokenValidationResult = decryptedJwt != null ? ValidateJWE(outerToken, decryptedJwt, validationParametersCopy) : ValidateJWS(token, validationParametersCopy); ;
+
+                            if (tokenValidationResult.IsValid) return tokenValidationResult;
+                        }
+                    }
+
+                    // If we were still unable to validate, attempt to refresh the configuration and validate using it.
+                    validationParametersCopy.ConfigurationManager.RequestRefresh();
+                    var lastConfig = validationParametersCopy.Configuration;
+                    validationParametersCopy.Configuration = validationParametersCopy.ConfigurationManager.GetBaseConfigurationAsync(CancellationToken.None).Result;
```

Comment:
"Sync over async" is always a code smell / bad pattern and it lead to thread-pool starvation and deadlocks. Ideally we would have a complete async stack from top-down. 
When it's time to refresh, we might have multiple threads waiting on this lock. As we use "sync over async", number of threads used is actually double.

Reply:
As a side not, using _GetAwaiter().GetResult()_ might be preferable as _.Result_ wraps the exception and throws AggregateException.